---
title: "Gene Set Annotation"
author: "Rebecca Lowdon"
date: "May 20, 2016"
output: html_document
---

Gene set annotations for NCVs in ChromHMM-18 regulatory states (1,2,7,8,14,15).

```{r}
setwd("~/cancer_annotation/KEGG/")
library(AnnotationDbi)
library(org.Hs.eg.db)
library(dplyr)
```

* Read in SNV-containing REs-to-promoters list and KEGG patwhay gene list filenames.

```{r, echo=FALSE}
# Genes associated w/ ChromHMM18 regulatory state NCVs
# Gene assignments made using GREAT
# gs <- read.table("GREAT-assocGenes_RM78-EIDs_allRegulStates_livWgsNcvs.txt",sep="\t",
#                  colClasses = c("factor","NULL"), skip = 1 )

# Gene assignments from master annotation script
gs <- read.table("RM78-EIDs_WGS_allRegulStates.SNVs_with_GeneID.tsv",sep = "\t",header = T)

# Get SYMBOL names
gs_names <- AnnotationDbi::select(x = org.Hs.eg.db, keys = as.character(gs$genes.gene_id), columns = "SYMBOL", keytype = "ENTREZID")
  head(gs_names)
  
# GSEA MDSig and KEGG pathway gene lists
annot_gene_sets <- read.table("genesets",header=F)
  head(annot_gene_sets)
```

* Function to calcuate and report overlap for each set.

```{r}
count_gene_set_overlap <- function( fl,x ){
  
  # test case:
#   d <- read.table(as.character("geneset-95.txt"), skip = 2, col.names="SYMBOL") 
#   pref <- read.table( as.character("geneset-95.txt"), nrows = 1)

  tab <- matrix( nrow = length(fl$V1), ncol = 3)
                     
  for( i in 1:length(fl$V1)){
    pref <- read.table( as.character(fl$V1[i]), nrows = 1)
    name <- pref$V1
    d <- read.table( as.character(fl$V1[i]), skip = 2, col.names="SYMBOL")
    overlap <- data.frame(unique(merge(x = x, y = d, by = "SYMBOL")))
    tab[i,1] <- paste(name)
    tab[i,2] <- length(overlap$SYMBOL)
    tab[i,3] <- length(d$SYMBOL)
  }
  return( tab )
}

df <- data.frame(count_gene_set_overlap( annot_gene_sets,gs_names ))
colnames(df) <- c("Pathway","numNcvs","numGenes")
  head(df)
```


## Monte Carlo permutations

* Function takes the list of all genes and creates a simulated gene list (`s_symbols`).
* Simulated gene list then run though the `count_gene_set_overlap()` function as above for the observed data.
* Retun a data frame with the results of 1000 simulations.

* First create gene list:

```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- keepStandardChromosomes(TxDb.Hsapiens.UCSC.hg19.knownGene)
  seqlevels(txdb)
  unique(genome(txdb))
genes <- GenomicFeatures::genes(txdb)
  head(genes)
  
gl <- data.frame(names(genes))
  head(gl)  
```


```{r}
monte_carlo <- function( sample_list, gene_annotations ){
 
  all_sdf <- data.frame()
 
  for( i in 1:10){
    s <- sample(sample_list, size = 690,replace = T)
    s_symbols <- AnnotationDbi::select(org.Hs.eg.db, keys = as.character(s),
                                      keytype = "ENTREZID", columns = "SYMBOL")
    sdf <- data.frame(count_gene_set_overlap( gene_annotations,s_symbols ))
    colnames(sdf) <- c("Pathway","numNcvs","numGenes")
    sdf$iteration <- i
 
    all_sdf <- rbind(all_sdf,sdf)
  }
  return(all_sdf)
}
 
mc <- monte_carlo( gl$names.genes., annot_gene_sets)
  glimpse(mc)

write.table("mc_10.tsv",quote = F, row.names = F, col.names = T, sep="\t")
```

* Read in saved result.

```{r}
mc <- read.table("mc_1000.tsv",sep="\t", header = T)
  glimpse(mc)
```


# Statistical test

* Function to find p-value for all pahtways.
* Returns a data frame

```{r}
pathways <- unique(mc$Pathway)
specify_decimal <- function(x, k) format(round(x, k), nsmall=k)
  
# Function for finding p-values for all pathways
find_p_value <- function( mc_data,pathway_list,obs_data ){
 
  pvals <- data.frame(matrix(ncol = 2))
  colnames(pvals) <- c("path","pval")
 
  for( i in 1:186 ){
    p <- dplyr::filter( mc_data, Pathway==paste(pathway_list[i]) )
    p$numNcvs <- as.numeric(paste0(p$numNcvs))
    p$numGenes <- as.numeric(paste0(p$numGenes))
 
    nk <- as.numeric(paste0(obs_data$numNcvs[ obs_data$Pathway==paste(pathway_list[i]) ]))
    sum_ns <- sum(p$numNcvs >= nk)
    pv <- sum_ns / 1000
 
    # hist(p$numNcvs,breaks = 15, xlim = c(0,nk+5));abline(v=nk, col="red")
    a <- data.frame( cbind(paste(pathway_list[i]),specify_decimal( pv,3 )) )
    colnames(a) <- c("path","pval")
    pvals <- rbind( pvals,a )
  }
  return( pvals )
}
 
mc_with_pval <- find_p_value( mc_data = mc, pathway_list = pathways, obs_data = df )
 
# Clean data
mc_with_pval <- filter(mc_with_pval, !is.na(path))
mc_with_pval$pval <- as.numeric(paste0(mc_with_pval$pval))
  head(mc_with_pval)

hist(mc_with_pval$pval, breaks = 20) ; abline(v=0.05, col="red")
```


## Inspect specific pathways

```{r}
library(dplyr)
library(ggplot2)

p <- dplyr::filter(mc, Pathway=="KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION")
  p$numNcvs <- as.numeric(paste0(p$numNcvs))
  p$numGenes <- as.numeric(paste0(p$numGenes))
  glimpse(p)  
 
nk <- as.numeric(paste0(df$numNcvs[df$Pathway=="KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION"]))
 
hist(p$numNcvs,breaks = 10, xlim = c(0,nk+5));abline(v=nk, col="red")
```


## Filter significant pathways
* Histogram of gene hits across 1000 smiulations

```{r}
pval_lt_0.05_with_numGenes <- filter(mc_with_pval, pval <=0.05)
  pval_lt_0.05_with_numGenes

write.table("pval_lt_0.05_with_numGenes.tsv", quote = F, row.names = F, col.names = T, sep = "\t")
```

* Inspect specific gene hits

```{r}
# KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION geneset-12.txt
# KEGG_DRUG_METABOLISM_CYTOCHROME_P450 geneset-43.txt

d <- read.table( file = "geneset-43.txt", skip = 2, col.names="SYMBOL")
overlap <- data.frame(unique(merge(x = gs_names, y = d, by = "SYMBOL")))
overlap

s <- filter(gs, grepl( "1544|1571|1576|2329|2940|54659|54657|54579", genes.gene_id, perl = T))
s$genes.start - s$re.start

```




* First round plots

```{r}
labs <- read.delim("geneset_groups",header=F)
colnames(labs) <- c("Pathway","set")
  head(labs)

library(ggplot2)

df$Pathway <- factor(df$Pathway)
df$numNcvs <- as.numeric(paste0(df$numNcvs))
df$numGenes <- as.numeric(paste0(df$numGenes))
  head(df)

df_path <- dplyr::left_join(df, labs, by.x=Pathway)
  head(df_path)

g <- filter(df_path, set=="cb") %>%
  ggplot(., aes(x=reorder(Pathway,numNcvs),y=numNcvs/numGenes,fill=set)) + geom_bar(stat = "identity", fill="yellow") + coord_flip()

g <- g + ylab("Number regulatory NCVs / number total genes annotated in pathway") +
  theme(axis.text.x=element_text(size=16), axis.title.x=element_text(size=16),
        axis.text.y=element_text(size=16), axis.title.y=element_blank(),
        legend.position="none")

png("KEGG_metabolism_livWgsNcvs_normalized.png",width = 1000, height = 1200)
g
dev.off();

```

